lib/polygeist/Passes/ConvertParallelToGPU.cpp

```diff
@@ -0,0 +1,1360 @@
+//===- ConvertParallelToGPU.cpp - Remove unused private functions ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "PassDetails.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/NVVMDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/IntegerSet.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "polygeist/BarrierUtils.h"
+#include "polygeist/Ops.h"
+#include "polygeist/Passes/Passes.h"
+#include "polygeist/Passes/Utils.h"
+
+// TODO when we add other backends, we would need to to add an argument to the
+// pass which one we are compiling to to provide the appropriate error id
+#if POLYGEIST_ENABLE_CUDA
+#include <cuda.h>
+#else
+#define CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES 701
+#endif
+
+using namespace mlir;
+using namespace polygeist;
+
+#define DEBUG_TYPE "convert-parallel-to-gpu"
+#define DBGS() ::llvm::dbgs() << "[" DEBUG_TYPE ":" << PATTERN << "] "
+
+namespace {
+
+template <typename T>
+bool hasEffect(SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  for (auto it : effects)
+    if (isa<T>(it.getEffect()))
+      return true;
+  return false;
+}
+
+template <int S = 3> SmallVector<Value, S> getUpperBounds(scf::ParallelOp pop) {
+  SmallVector<Value, S> bounds;
+  for (auto bound : pop.getUpperBound()) {
+    bounds.push_back(bound);
+  }
+  return bounds;
+}
+
+void insertReturn(PatternRewriter &rewriter, func::FuncOp f) {
+  rewriter.create<func::ReturnOp>(rewriter.getUnknownLoc());
+}
+void insertReturn(PatternRewriter &rewriter, LLVM::LLVMFuncOp f) {
+  rewriter.create<LLVM::ReturnOp>(rewriter.getUnknownLoc(),
+                                  std::vector<Value>{});
+}
+
+/// a / 2 < prevPowerOf2(a) <= a
+unsigned prevPowerOf2(unsigned a) {
```

Comment:
Can you use the existing LLVM versions? (https://llvm.org/doxygen/MathExtras_8h.html)

Reply:
done